src/test/java/io/netty/incubator/codec/http3/QpackEncoderDecoderTest.java

```diff
@@ -17,64 +17,194 @@
         out.writeByte(0);
 
         try {
-            decoder.decode(out, new Http3HeadersSink(decHeaders, 1024, false));
+            decode(out, decHeaders);
             fail();
         } catch (QpackException exception) {
             // expected
         }
-        out.release();
+    }
+
+    private void encode(ByteBuf buf, Http3Headers headers) {
+        encoder.encodeHeaders(DEFAULT, 1, encoderStream, buf, headers);
+    }
+
+    private void decode(ByteBuf buf, Http3Headers headers) throws QpackException {
+        decoder.decode(decoderStream, DEFAULT, 1, buf, buf.readableBytes(),
+                new Http3HeadersSink(headers, 1024, false), whenDecoded);
+    }
+
+    private void verifyDecodedHeader(CharSequence name, CharSequence value) {
+        assertThat(decHeaders.get(name), is(new AsciiString(value)));
+    }
+
+    private void setup(long dynamicTableSize, int maxBlockedStreams) throws Exception {
+        DefaultHttp3SettingsFrame localSettings = new DefaultHttp3SettingsFrame();
+        localSettings.put(HTTP3_SETTINGS_QPACK_MAX_TABLE_CAPACITY, dynamicTableSize);
+        localSettings.put(HTTP3_SETTINGS_QPACK_BLOCKED_STREAMS, (long) maxBlockedStreams);
+        when(syncStrategy.entryAdded(anyLong())).thenAnswer(__ -> stateSyncStrategyAckNextInsert);
+        final QpackHuffmanDecoder huffmanDecoder = new QpackHuffmanDecoder();
+        decoder = new QpackDecoder(huffmanDecoder, localSettings, decDynamicTable, syncStrategy);
+        encoder = new QpackEncoder(encDynamicTable);
+        EmbeddedQuicChannel parent = new EmbeddedQuicChannel();
+        QpackEncoderHandler encoderHandler = new QpackEncoderHandler(dynamicTableSize, huffmanDecoder, decoder);
+        encoderStream = (EmbeddedQuicStreamChannel) parent.createStream(UNIDIRECTIONAL,
+                new ChannelOutboundHandlerAdapter() {
+                    @Override
+                    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+                        if (msg instanceof ByteBuf) {
+                            encoderHandler.channelRead(ctx, msg);
+                        } else {
+                            super.write(ctx, msg, promise);
+                        }
+                    }
+        }).get();
+        encoder.settings(encoderStream, encoderStream.alloc(), localSettings);
+        QpackDecoderHandler decoderHandler = new QpackDecoderHandler(encoder);
+        decoderStream = (EmbeddedQuicStreamChannel) parent.createStream(UNIDIRECTIONAL,
+                new ChannelOutboundHandlerAdapter() {
+                    @Override
+                    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+                        if (msg instanceof ByteBuf) {
+                            decoderHandler.channelRead(ctx, msg);
+                        } else {
+                            super.write(ctx, msg, promise);
+                        }
```

Comment:
as the implementation is basically the same as above you may create your own handler class that you can-reuse.